velox/exec/tests/SpillerTest.cpp

```diff
@@ -22,183 +22,323 @@
 
 using namespace facebook::velox;
 using namespace facebook::velox::exec;
+using namespace facebook::velox::exec::test;
 using namespace facebook::velox::common::testutil;
+using facebook::velox::filesystems::FileSystem;
+
+struct TestParam {
+  Spiller::Type type;
+  // Specifies the spill executor pool size. If the size is zero, then spill
+  // write path is executed inline with spiller control code path.
+  int poolSize;
+
+  TestParam(Spiller::Type _type, int _poolSize)
+      : type(_type), poolSize(_poolSize) {}
+};
+
+std::vector<TestParam> getTestParams() {
+  std::vector<TestParam> params;
+  for (int i = 0; i < Spiller::kNumTypes; ++i) {
+    const auto type = static_cast<Spiller::Type>(i);
+    // TODO: add to test kHashJoin after it is supported.
+    if (type == Spiller::Type::kHashJoin) {
+      continue;
+    }
+    for (int poolSize : {0, 8}) {
+      params.emplace_back(type, poolSize);
+    }
+  }
+  return params;
+}
 
 class SpillerTest : public exec::test::RowContainerTestBase,
-                    public testing::WithParamInterface<int> {
+                    public testing::WithParamInterface<TestParam> {
  public:
   static void SetUpTestCase() {
     TestValue::enable();
   }
 
+  SpillerTest()
+      : type_(GetParam().type),
+        executorPoolSize_(GetParam().poolSize),
+        hashBits_(0, type_ == Spiller::Type::kOrderBy ? 0 : 2),
+        numPartitions_(hashBits_.numPartitions()) {}
+
+  void SetUp() override {
+    RowContainerTestBase::SetUp();
+    tempDirPath_ = exec::test::TempDirectoryPath::create();
+    fs_ = filesystems::getFileSystem(tempDirPath_->path, nullptr);
+  }
+
  protected:
   void testSpill(int32_t spillPct, bool makeError = false) {
-    constexpr int32_t kNumRows = 100000;
-    std::vector<char*> rows(kNumRows);
-    RowVectorPtr batch;
-    auto data = makeSpillData(kNumRows, rows, batch);
-    std::vector<uint64_t> hashes(kNumRows);
-    auto keys = data->keyTypes();
+    constexpr int32_t kNumRows = 100'000;
+    std::shared_ptr<const RowType> rowType = ROW({
+        {"bool_val", BOOLEAN()},
+        {"tiny_val", TINYINT()},
+        {"small_val", SMALLINT()},
+        {"int_val", INTEGER()},
+        {"long_val", BIGINT()},
+        {"ordinal", BIGINT()},
+        {"float_val", REAL()},
+        {"double_val", DOUBLE()},
+        {"string_val", VARCHAR()},
+        {"array_val", ARRAY(VARCHAR())},
+        {"struct_val", ROW({{"s_int", INTEGER()}, {"s_array", ARRAY(REAL())}})},
+        {"map_val",
+         MAP(VARCHAR(),
+             MAP(BIGINT(),
+                 ROW({{"s2_int", INTEGER()}, {"s2_string", VARCHAR()}})))},
+    });
+    setupSpillData(rowType, 6, kNumRows, [&](RowVectorPtr rows) {
+      // Set ordinal so that the sorted order is unambiguous
```

Comment:
Should we have a test with duplicate sorting keys as well? I believe this case will happen in order-by spilling.

Reply:
Good point. I leave a TODO to add duplicate sort keys for SpillTest and SpillerTest in a separate followup. Thanks!